src/gDP.cpp

```diff
@@ -792,6 +797,147 @@ void gDPSetScissor(u32 mode, s16 xh, s16 yh, s16 xl, s16 yl)
 #endif
 }
 
+// This performs the same thing action as gDPFillRectangle but explicitly specifying what to overwrite
+void gDPMemset(u32 value, u32 addr, u32 length)
+{
+	// TODO: Can this be different?
+	const u32 screenWidth = VI.width;
+	const u32 screenHeight = VI.height;
+
+	GraphicsDrawer& drawer = dwnd().getDrawer();
+	u32 fillColor = value | (value << 16);
+	const u32 depthImageStart = gDP.depthImageAddress;
+	const u32 depthImageEnd = gDP.depthImageAddress + screenWidth * screenHeight * 2;
+	if (depthImageStart <= addr && addr < depthImageEnd) {
+		// ZB clear
+		gDP.rectColor.r = _FIXED2FLOATCOLOR(_SHIFTR(fillColor, 11, 5), 5);
+		gDP.rectColor.g = _FIXED2FLOATCOLOR(_SHIFTR(fillColor, 6, 5), 5);
+		gDP.rectColor.b = _FIXED2FLOATCOLOR(_SHIFTR(fillColor, 1, 5), 5);
+		gDP.rectColor.a = _SHIFTR(fillColor, 0, 1);
+
+		// HACK: this usually replaces gDPSetColorImage for zb so save zb
+		frameBufferList().saveBuffer(gDP.depthImageAddress, (u16)G_IM_FMT_RGBA, (u16)G_IM_SIZ_16b, (u16)screenWidth, false);
+
+		if (config.generalEmulation.enableFragmentDepthWrite == 0)
+			drawer.clearDepthBuffer();
+		else
+			depthBufferList().setCleared(true);
+
+		// Deduce ulx, uly, lrx, lry from addr, size, and colorImageStart
+		// Currently I am assuming a simple letterbox case - top and bottom are not cleared
+		const u32 lineSize = screenWidth * 2;
+		const u32 ulx = 0;
+		const u32 uly = (addr - depthImageStart) / lineSize;
+		const u32 lrx = screenWidth;
+		const u32 lry = uly + length / lineSize;
+
+		if (config.generalEmulation.enableFragmentDepthWrite != 0) {
+			// Pretend that we are drawing the rectangle over ZB with fill mode
+			ValueKeeper<u32> backupColorImageAddress(gDP.colorImage.address, gDP.depthImageAddress);
+			ValueKeeper<u32> backupWidth(gDP.colorImage.width, screenWidth);
+			const unsigned int backupDepthSource = gDP.otherMode.depthSource;
+			const unsigned int backupCycleType = gDP.otherMode.cycleType;
+
+			gDP.otherMode.cycleType = G_CYC_FILL;
+
+			drawer.drawRect(ulx, uly, lrx, lry);
+			frameBufferList().setBufferChanged(f32(lry));
+
+			gDP.otherMode.depthSource = backupDepthSource;
+			gDP.otherMode.cycleType = backupCycleType;
```

Comment:
We still can use ValueKeeper there and make code a bit shorter and safe:
```suggestion
			gDPInfo::OtherMode otherMode = gDP.otherMode;
			otherMode.cycleType = G_CYC_FILL;
			ValueKeeper<gDPInfo::OtherMode> backupOtherMode(gDP.otherMode, otherMode);
			drawer.drawRect(ulx, uly, lrx, lry);
			frameBufferList().setBufferChanged(f32(lry));
```